### DISCUSSION TITLE
nixos-with-flakes/downgrade-or-upgrade-packages

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#136] [Announcements]
# nixos-with-flakes/downgrade-or-upgrade-packages

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/nixos-with-flakes/downgrade-or-upgrade-packages

**Comment by parmigggiana** [comment]
This might be obvious but it wasn't for me, so I'll leave it here in case anyone else encounters the same problem;
The code shown in this specific example doesn't immediately work for me. Not sure why, but it seems like home-manager doesn't pick up on the specified specialArgs. I could get it to work by using `home-manager.extraSpecialArgs` instead. 
For reference, this is my flake.nix
```nix
{
  description = "Main NixOS flake";

  inputs = {
    # NixOS official package source, using the nixos-23.11 branch here
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.11";
    nixpkgs-unstable.url = "github:NixOS/nixpkgs/nixos-unstable";
    nixos-hardware.url = "github:NixOS/nixos-hardware/master";
    home-manager = {
      url = "github:nix-community/home-manager/release-23.11";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    # fw-fanctrl.url = "github:mdvmeijer/fw-fanctrl-nix";
  };

  outputs = inputs@{ 
    self, 
    nixpkgs, 
    nixpkgs-unstable, 
    nixos-hardware, 
    home-manager, 
    ...
  }: {
    nixosConfigurations.<hostname> = nixpkgs.lib.nixosSystem rec {
      system = "x86_64-linux";
  
      modules = [
        # Hardware definition from https://github.com/NixOS/nixos-hardware/blob/master/flake.nix
        nixos-hardware.nixosModules.framework-12th-gen-intel
        ./default.nix

        home-manager.nixosModules.home-manager
        {
          home-manager.useGlobalPkgs = true;
          home-manager.useUserPackages = true;

          home-manager.users.<user> = import ./users/<user>/home.nix;

          home-manager.extraSpecialArgs = {
            pkgs-unstable = import nixpkgs-unstable {
              inherit system;
              config.allowUnfree = true;
            };
          };
        }
      ];
    };
  };
}
```

**Reply by callum-albert** [reply]
非常感谢你！我今天也遇到了同样的问题-Thank you so much! I had the same problem today

**Reply by AntonOfTheWoods** [reply]
Me too... maybe more recent versions require the `extraSpecialArgs` parameter?

**Comment by TheNeekOfficial** [comment]
How would one go about setting it with the home.packages way but with 2 different branches. ie. smth akin to this:

  home.packages = with pkgs; [
    ... # Insert packages here
  ];
  
  home.packages = with pkgs-stable; [
    ... # Insert packages here
  ];

I'm curious as to whether or not this is possible as it currently pops up as an error on my end and i understand why it doesnt work, just curious if there is a way to do it.
If anyone has an answer, thanks, otherwise i'll stick to Overlays if I ever need a more stable version of a package.

**Reply by ryan4yin** [reply]
This should work:

```nix
  home.packages =
    (with pkgs; [
      # ... # Insert packages here
    ])
    ++ (with pkgs-stable; [
      # ... # Insert packages here
    ]);
```

Alternatively, you can use this:

```nix
  home.packages = with pkgs; [
    # ... # Insert packages here

    # unstable packages
    pkgs-stable.a
    pkgs-stable.b
    pkgs-stable.c
  ];
```

**Reply by TheNeekOfficial** [reply]
oh shoot thank you! thanks for the quick reply too! v appreciated as a person just diving head first into Nix and trying to work it all out after only coding for like ~8 months and using Linux for 1 lmao. thanks

**Comment by thein3rovert** [comment]
So if i do this:     # Default to the nixos-unstable branch
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";

    # Latest stable branch of nixpkgs, used for version rollback
    # The current latest version is 24.11
    nixpkgs-stable.url = "github:nixos/nixpkgs/nixos-24.11";
and i run nix flake update..will it update my nixos version from 24.11 to the latest version..

I recent;y has an issue where it updated my nixos to the warbler version and it has breaking changes. I want to avoid that.

How do i update my flake without updating my nixos version.

**Comment by hf02** [comment]
I was looking into how to accomplish this with home-manager alongside the regular Nix config. Tried this first... it works, but it's not very elegant as there's a lot of repeating

```nix
  outputs = inputs@{
    self,
    nixpkgs,
    nixpkgs-unstable,
    ...
  }: {
    nixosConfigurations = {
      my-nixos = nixpkgs.lib.nixosSystem {
        system = "x86_64-linux";

        specialArgs = {
          pkgs-unstable = import nixpkgs-unstable {
            system = "x86_64-linux";
            config.allowUnfree = true;
          };
        };

        modules = [
          ./hosts/my-nixos

          home-manager.nixosModules.home-manager {
            home-manager.useGlobalPkgs = true;
            home-manager.useUserPackages = true;

            home-manager.users.you = import ./home.nix;

            home-manager.extraSpecialArgs = {
              pkgs-unstable = import nixpkgs-unstable {
                system = "x86_64-linux";
                config.allowUnfree = true;
              };
            };
          };
        ];
      };
    };
  };
```

Came up with this. Define pkgs-unstable beforehand using `let ... in`, and just have specialArgs and extraSpecialArgs inherit from it.

```nix
  outputs = inputs@{
    self,
    nixpkgs,
    nixpkgs-unstable,
    ...
  }:
  let 
    system = "x86_64-linux";
    pkgs-unstable = import nixpkgs-unstable {
      system = system;
      config.allowUnfree = true;
    };
  in
  {
    nixosConfigurations = {
      my-nixos = nixpkgs.lib.nixosSystem {
        system = "x86_64-linux";

        specialArgs = {
          inherit pkgs-unstable;
        };

        modules = [
          ./hosts/my-nixos

          home-manager.nixosModules.home-manager {
            home-manager.useGlobalPkgs = true;
            home-manager.useUserPackages = true;

            home-manager.users.you = import ./home.nix;

            home-manager.extraSpecialArgs = {
              inherit pkgs-unstable;
            };
          };
        ];
      };
    };
  };
```
